arch/sparc/src/s698pm/s698pm-irq.c

```diff
@@ -0,0 +1,533 @@
+
+/****************************************************************************
+ * Name: up_irqinitialize
+ ****************************************************************************/
+
+void up_irqinitialize(void)
+{
+  int irq = 0;
+
+  for (irq = 0; irq < NR_IRQS; irq++)
+    {
+      g_irqmap[irq] = IRQ_UNMAPPED;
+    }
+
+  /* Initialize CPU interrupts */
+
+  s698pm_cpuint_initialize();
+
+  /* Set interrupts priority */
+
+  for (irq = 0x11; irq < 0x20; irq++)
+    {
+      /* Set all interrupts to the default (low) priority */
+
+      (void)up_prioritize_irq(irq, 0);
+    }
+
+  /* Attach software interrupts */
+
+  irq_attach(S698PM_IRQ_SW_SYSCALL_TA0, up_swint0, NULL);
+  irq_attach(S698PM_IRQ_SW_SYSCALL_TA8, up_swint1, NULL);
+
+  /* And finally, enable cpu interrupts */
+
+#ifndef CONFIG_SUPPRESS_INTERRUPTS
+  up_irq_enable();
+#endif
+}
+
+/****************************************************************************
+ * Name:  s698pm_cpuint_initialize
+ *
+ * Description:
+ *   Initialize CPU interrupts
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; A negated errno value is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+int s698pm_cpuint_initialize(void)
+{
+  uintptr_t regaddr;
+  int cpu = 0;
+
+#ifdef CONFIG_SMP
+  /* Which CPU are we initializing */
+
+  cpu = up_cpu_index();
+  DEBUGASSERT(cpu >= 0 && cpu < CONFIG_SMP_NCPUS);
+#endif
+
+  /* get Interrupt_Mask reg address of cpu */
+
+  regaddr = S698PM_IRQREG_P0_MASK + 4 * cpu;
+
+  /* Disable all CPU interrupts on this CPU */
+
+  putreg32(0, regaddr);
+
+#if defined CONFIG_SMP
+  /* Attach IPI interrupts */
+
+  irq_attach(S698PM_IPI_IRQ, s698pm_pause_handler, NULL);
+
+  (void)s698pm_setup_irq(cpu, S698PM_IPI_IRQ, 0);
+
+  /* And enable the IPI interrupt */
+
+  up_enable_irq(S698PM_IPI_IRQ);
+#endif
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name:  s698pm_setup_irq
+ *
+ * Description:
+ *   This function sets up the IRQ. It allocates a CPU interrupt of the given
+ *   priority andattaches it to the given irq.
+ *
+ * Input Parameters:
+ *   cpu      - The CPU to receive the interrupt 0~3
+ *   irq      - The irq number from irq.h to be assigned to a EXT interrupt.
+ *   priority - Interrupt's priority (0~1).
+ *
+ * Returned Value:
+ *   The allocated CPU interrupt on success, a negated errno value on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int s698pm_setup_irq(int cpu, int irq, int priority)
+{
+  irqstate_t irqstate;
+  int cpuint;
+
+  irqstate = enter_critical_section();
+
+  if (irq >= S698PM_IRQ_FIRST_INT && irq <= S698PM_IRQ_LAST_INT)
+    {
+      cpuint = irq - S698PM_IRQ_FIRST_INT + 1;
+    }
+  else if (irq > S698PM_IRQ_LAST && irq < NR_IRQS)
+    {
+      cpuint = irq - 240;
```

Comment:
What is 240? Please avoid hard-coded value that could be difficult to track later